internal/core/src/index/TextMatchIndex.cpp

```diff
@@ -0,0 +1,168 @@
+        LOG_INFO(
+            "text index {} already exists, which should happen in loading "
+            "progress",
+            path_);
+        wrapper_ = std::make_shared<TantivyIndexWrapper>(path_.c_str());
+    } else {
+        std::string field_name =
+            std::to_string(disk_file_manager_->GetFieldDataMeta().field_id);
+        wrapper_ = std::make_shared<TantivyIndexWrapper>(
+            field_name.c_str(), false, path_.c_str());
+    }
+}
+
+BinarySet
+TextMatchIndex::Upload(const Config& config) {
+    finish();
+
+    boost::filesystem::path p(path_);
+    boost::filesystem::directory_iterator end_iter;
+
+    for (boost::filesystem::directory_iterator iter(p); iter != end_iter;
+         iter++) {
+        if (boost::filesystem::is_directory(*iter)) {
+            LOG_WARN("{} is a directory", iter->path().string());
+        } else {
+            LOG_INFO("trying to add text log: {}", iter->path().string());
+            AssertInfo(disk_file_manager_->AddTextLog(iter->path().string()),
+                       "failed to add text log: {}",
+                       iter->path().string());
+            LOG_INFO("text log: {} added", iter->path().string());
+        }
+    }
+
+    BinarySet ret;
+
+    auto remote_paths_to_size = disk_file_manager_->GetRemotePathsToFileSize();
+    for (auto& file : remote_paths_to_size) {
+        ret.Append(file.first, nullptr, file.second);
+    }
+
+    return ret;
+}
+
+void
+TextMatchIndex::Load(const Config& config) {
+    auto index_files =
+        GetValueFromConfig<std::vector<std::string>>(config, "index_files");
+    AssertInfo(index_files.has_value(),
+               "index file paths is empty when load text log index");
+    auto prefix = disk_file_manager_->GetLocalTextIndexPrefix();
+    disk_file_manager_->CacheTextLogToDisk(index_files.value());
+    wrapper_ = std::make_shared<TantivyIndexWrapper>(prefix.c_str());
+}
+
+void
+TextMatchIndex::AddText(const std::string& text, int64_t offset) {
+    AddTexts(1, &text, offset);
+}
+
+void
+TextMatchIndex::AddTexts(size_t n,
+                         const std::string* texts,
+                         int64_t offset_begin) {
+    wrapper_->add_data(texts, n, offset_begin);
+    if (shouldTriggerCommit()) {
+        Commit();
+    }
+}
+
+void
+TextMatchIndex::Finish() {
+    finish();
+}
+
+bool
+TextMatchIndex::shouldTriggerCommit() {
+    auto span = (std::chrono::duration<double, std::milli>(
+                     stdclock::now() - last_commit_time_.load()))
+                    .count();
+    return span > commit_interval_in_ms_;
+}
+
+void
+TextMatchIndex::Commit() {
+    wrapper_->commit();
+    last_commit_time_.store(stdclock::now());
+}
+
+void
+TextMatchIndex::Reload() {
+    wrapper_->reload();
+}
+
+void
+TextMatchIndex::CreateReader() {
+    wrapper_->create_reader();
+}
+
+TargetBitmap
+TextMatchIndex::MatchQuery(const std::string& query) {
+    if (shouldTriggerCommit()) {
```

Comment:
what happens if concurrent Commit and Reload is triggered?

Reply:
try lock optimization can be applied.